### ISSUE TITLE
Cache refresh broken when cache_resource_clear_partial is enabled

### ISSUE THREAD

**Issue by nsuomine** [modxcms/revolution#16343]
### Summary
I upgraded my site to MODX 3.0.0 and then to 3.0.1. When I select Manage - Clear Cache, it looks like it is cleared normally. But all the data stays in the core/cache folder. The only way I can clear the cache is to remove the cache directory.

### Step to reproduce
1. Enable `cache_resource_clear_partial` system setting.
2. Clear cache by _Manage - Clear Cache_ or edit a resource and keep the Empty Cache setting on.
3. View the resource. MODX returns the old version even though the resource was updated in the database. The old version still exists in the cache folder.

### Observed behavior
MODX returns old versions until I remove the cache folder.

### Expected behavior
MODX should return the updated versions.

### Environment
MODX 3.0.1

### Error Log
I get these lines to the error log while clearing
.../core/vendor/xpdo/xpdo/src/xPDO/Cache/xPDOCacheManager.php : 498) PHP warning: unlink(.../core/cache/db/objects/MODX/Revolution/modDeprecatedCall/9cbe0f3759eba6bd21f57cb9c6e30fed.cache.php): No such file or directory

.../core/vendor/xpdo/xpdo/src/xPDO/Cache/xPDOCacheManager.php : 506) PHP warning: closedir(): supplied resource is not a valid Directory resource

**Comment by nsuomine** [comment]
I did some debugging:

XPDOFileCache tries to delete the file .../public_html/dev/core/cache/resource/web.cache.php which naturally fails since it should delete the contents of the web folder.

The problem is this line:
` if ($this->getOption(xPDO::OPT_CACHE_MULTIPLE_OBJECT_DELETE, $options, false)) {`

It returns false but it should return true. xPDO::OPT_CACHE_MULTIPLE_OBJECT_DELETE option does not exist so it sticks to the default value. I don't have enough skills to find out why this option is missing.

**Comment by opengeek** [comment]
I cannot reproduce any issue with clearing the cache using either method. When I update a resource, the cache entries for that resource are removed and regenerated on next load. The cache directories are not removed by the refresh() method that is called when you "clear" the cache. Only the files within it are deleted. You might want to explore what problem you are encountering in the forums or our Slack community.

**Comment by nsuomine** [comment]
I have two MODX3 sites. The other one clears the cache normally but the other one does not.

The folder .../public_html/dev/core/cache/resource/web/ is not emptied. Instead, the code tries to unlink file .../public_html/dev/core/cache/resource/web.cache.php (see my first (edited) comment). So it assumes the folder is a file.

**Comment by opengeek** [comment]
Then you likely have some kind of extension or plugin that may be causing the problem on the one site.

**Comment by nsuomine** [comment]
But is it normal that this line returns false?

`if ($this->getOption(xPDO::OPT_CACHE_MULTIPLE_OBJECT_DELETE, $options, false)) {`

Because it returns false, the code does not try to empty the folder. It returns false because xPDO::OPT_CACHE_MULTIPLE_OBJECT_DELETE does not exist.

**Comment by opengeek** [comment]
Yes. That option is only used in particular circumstances.

**Comment by nsuomine** [comment]
Ok. So the line should return false. The code jumps then here:

```
$cacheKey= $this->getCacheKey($key, $options);
        if (file_exists($cacheKey)) {
            $deleted= @ unlink($cacheKey);
        }
```

It does not try to empty the resource folder but only unlink a file.  

This part inside the if condition is not executed:

```
  if ($this->getOption(xPDO::OPT_CACHE_MULTIPLE_OBJECT_DELETE, $options, false)) {
            $cacheKey= $this->getCacheKey($key, array_merge($options, array('cache_ext' => '')));
            if (file_exists($cacheKey) && is_dir($cacheKey)) {
                $results = $this->xpdo->cacheManager->deleteTree($cacheKey, array_merge(array('deleteTop' => false, 'skipDirs' => false, 'extensions' => array('.cache.php')), $options));
                if ($results !== false) {
                    $deleted = true;
                }
            }
        }
```

But deletetree is inside that condition.

**Comment by nsuomine** [comment]
I compared my working site and non-working site and found a difference.

When the setting cache_resource_clear_partial is on, cache\resource folder is not emptied. If I turn it off, clearing the folder works fine.

**Comment by nsuomine** [comment]
Yes, the setting solved my problem. But shouldn't resource cache clearing be working also when the cache_resource_clear_partial  is on?

**Comment by opengeek** [comment]
AFAIK, the whole point of that setting is to prevent clearing the resource cache other than the specific resource you are editing. But I honestly don't know anything about that feature and will have to defer to someone who does. Regardless, if there are issues with that feature, we'll need details relevant to that specific problem. Exploring this in the forums or Slack chat as I previously suggested would be the best way to start that process.

**Comment by nsuomine** [comment]
"When enabled, MODX refresh will only clear resource cache for the provided contexts."

And it does it in a foreach loop for provided contexts. But as I wrote before, at least in my situation, deletetree methods are not called and the folders under cache/resource are not emptied. This worked still in 2.8 version. When cache_resource_clear_partial  is off, clearing seems to happen in a different way and that is why it works.

And I have opened a discussion in Stack too and will continue there: https://stackoverflow.com/questions/75010452/cache-clearing-fails-in-modx3-0-2/75015833#75015833. Thanks for help.
